### PR TITLE
feat: add custom playback configuration

### PR DESCRIPTION
**Pull request by jonathanm-tkf**
This PR it's related to issue #9

thanks and regards

### CODE REVIEW

**Comment by pogist** [comment]
Hello @jonathanm-tkf, thank you very much for your contribution 😁. That's a nice feature to have indeed, but there are a few issues with your code I need you to tweak before I can approve:

- `PlaybackConfig` should be part of `PlayerConfig`:
```typescript
export interface PlayerConfig {
  // ...other properties.

  /**
   * Configures playback behavior. A default `PlaybackConfig` is set initially.
   */
  playbackConfig?: PlaybackConfig;
}
```

- Remove the `playbackConfig` parameter from `Player.constructor`, `usePlayer` and `initWithConfig` (native sources);
- Just handle playback config initialization on `JsonConverter` and `RCTConvert+BitmovinPlayer`, not directly on the `PlayerModule`;
- Add as many props from the native `PlaybackConfig` object as possible, not just `isAutoplayEnabled`. Other common props between android/iOS are `isMuted` and `isTimeShiftEnabled`;

**Comment by jonathanm-tkf** [comment]
Thanks, @pogist in principle I had thought about it that way but I didn't want to affect the current functionality, as soon as I have some time I'll try to implement it, if I have any doubt I'll comment on the issue :)

Regarding question

When implementing the solution in some cases in android we have the following detail, regarding the buffering? which occurs in a non-constant way, only happens on android as it works perfectly well on ios.

What would be the best way to solve this problem?

Thanks again

**Comment by jonathanm-tkf** [comment]
@pogist I don't understand what you mean by this, how would it be better? Could you give me an example?

- Just handle playback config initialization on JsonConverter and RCTConvert+BitmovinPlayer, not directly on the PlayerModule;

**Comment by jonathanm-tkf** [comment]
> @pogist I don't understand what you mean by this, how would it be better? Could you give me an example?
> 
> * Just handle playback config initialization on JsonConverter and RCTConvert+BitmovinPlayer, not directly on the PlayerModule;

Maybe you referred to something like that

```swift
  static func playerConfig(_ json: Any?) -> PlayerConfig? {
      let playerConfig = PlayerConfig()
      guard let json = json as? [String: Any?] else {
          return playerConfig
      }
      if let licenseKey = json["licenseKey"] as? String {
          playerConfig.key = licenseKey
      }
      // ...other logic
      if let playbackConfig = json["playbackConfig"] as? [String: Any?] {
          if let isAutoplayEnabled = playbackConfig["isAutoplayEnabled"] as? Bool {
              playerConfig.playbackConfig.isAutoplayEnabled = isAutoplayEnabled
          }
      }

      return playerConfig
  }

```

**Comment by jonathanm-tkf** [comment]
Hi @pogist 

I have changed the implementation could you check it and let me know if it is to your liking?

Greetings.

**Comment by jonathanm-tkf** [comment]
Thank you very much for the support @pogist, yes correct, and I totally agree I just added it as a new screen because I needed to test and not interfere with your original example, as soon as I get free I will be modifying and if you like maybe I can modify the readme with an example so it does not interfere with the BasicPlayback, that is what you think thank you very much.

**Comment by jonathanm-tkf** [comment]
Hi @pogist how are you?
I caught up with your recommendations and suggestions if you have time to review them, let me know if you agree or need any updates.
Thanks and regards.

File: README.md
Comment (pogist) [inline README.md:20]: Please, rename this section. It has some mistakes.

```suggestion
    - [Setting up the playback configuration](#setting-up-the-playback-configuration)
```

File: README.md
Comment (pogist) [inline README.md:23]: Those two topics are related to the `Setting up a license key` section, not the playback configuration.

File: README.md
Comment (pogist) [inline README.md:220]: Repeating the prop names in comments feels a little bit cumbersome here. You can just leave the description and default value.

```suggestion
    // Specifies whether the playback starts immediately after loading a source or not. Default is false.
```

File: README.md
Comment (pogist) [inline README.md:222]:
```suggestion
    // Specifies if playback starts muted. Default is false.
```

File: README.md
Comment (pogist) [inline README.md:224]:
```suggestion
    // Specifies if time shift for live streamings should be enabled. Default is true.
```

File: README.md
Comment (pogist) [inline README.md:226]:
```suggestion
    // Whether background playback is enabled or not. Default is false.
    // Only available for iOS.
```

File: README.md
Comment (pogist) [inline README.md:211]: You can remove this sub-section since it is the only one present under the main headline.

```suggestion
```

File: README.md
Comment (pogist) [inline README.md:209]:
```suggestion
If needed, the default player behavior can be configured through the `playbackConfig` key when initialized.
```

**Review by pogist — CHANGES_REQUESTED** [review]

File: src/player.ts
Comment (pogist) [inline src/player.ts:42]: Actually, we don't set a default value for `PlaybackConfig` so this doc comment could be a little misleading.

```suggestion
   * Configuration options for player's default playback behavior.
```

File: src/player.ts
Comment (pogist) [inline src/player.ts:48]:
```suggestion
 * Configuration options for player's default playback behavior.
```

File: src/player.ts
Comment (pogist) [inline src/player.ts:64]:
```suggestion
   * Whether the player starts playing automatically after loading a source or not. Default is `false`.
   *
   * @example
   * ```
   * const player = new Player({
   *   playbackConfig: {
   *     isAutoplayEnabled: true,
   *   },
   * });
   * ```
   */
```

File: src/player.ts
Comment (pogist) [inline src/player.ts:78]:
```suggestion
   * Whether the sound is muted on startup or not. Default value is `false`.
   *
   * @example
   * ```
   * const player = new Player({
   *   playbackConfig: {
   *     isMuted: true,
   *   },
   * });
   * ```
```

File: src/player.ts
Comment (pogist) [inline src/player.ts:93]:
```suggestion
   * Whether time shift / DVR for live streams is enabled or not. Default is `true`.
   *
   *  @example
   * ```
   * const player = new Player({
   *   playbackConfig: {
   *     isTimeShiftEnabled: false,
   *   },
   * });
   * ```
```

File: src/player.ts
Comment (pogist) [inline src/player.ts:127]:
```suggestion
   * Whether the picture-in-picture mode option is enabled for iOS or not. Default is `false`.
   *
   *  @example
   * ```
   * const player = new Player({
   *   playbackConfig: {
   *     isPictureInPictureEnabled: true,
   *   },
   * });
   * ```
```

**Review by pogist — CHANGES_REQUESTED** [review]

File: README.md
Comment (pogist) [reply README.md:23]: Please, move them back to the `Setting up a license key` section just the way it was.

**Comment by jonathanm-tkf** [comment]
Hi @pogist how are you?
I caught up with your recommendations and suggestions if you have time to review them, let me know if you agree or need any updates.
Pido perdon debido a que he estado bastante distante del proyecto pero fueron semanas bastante duras en el trabajo, mil disculpas
Thanks and regards.

**Comment by jonathanm-tkf** [comment]
Hi there @pogist 👋 

Some news with this integration?

Thanks for your support

**Review by pogist — APPROVED** [review]
Hi, @jonathanm-tkf 👋

Sorry I didn't get in touch with you in a while. I got caught up on work these last few weeks and totally forgot to leave my review the last time I checked, so I'm really sorry for it 🙏.

But I saw you applied the docs changes I asked and it seems really good to me, thanks for your effort 😁. We just need to wait for the native devs to review it now and then we'll be good to merge it. I believe we'll be able to include it on the next release already too. So again, thanks for your support and sorry about the delay.

**Comment by jonathanm-tkf** [comment]
Thanks, @pogist thanks for your support I really appreciate it!!!

File: src/player.ts
Comment (rolandkakonyi) [inline src/player.ts:32]: Please always use the existing documentation from the native SDKs to ensure consistency.

```suggestion
   * Configures playback behaviour. A default PlaybackConfig is set initially.
```

File: src/player.ts
Comment (rolandkakonyi) [inline src/player.ts:38]: Please use either the iOS or the Android SDKs documentation for public API documentation for this library

```suggestion
 * Configures the playback behaviour of the player.
```

File: src/player.ts
Comment (rolandkakonyi) [inline src/player.ts:80]: Please always use the native SDKs documentation as they provide detailed explanation for given configurations which might be crucial for their usage.

```suggestion
   * Whether background playback is enabled or not.
   * Default is `false`.
   * 
   * When set to `true`, playback is not automatically paused
   * anymore when the app moves to the background.
   * When set to `true`, also make sure to properly configure your app to allow
   * background playback.
   *
   * On tvOS, background playback is only supported for audio-only content.
```

File: android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt
Comment (matamegger) [inline android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt:39]: Please also apply @rolandkakonyi suggestion about the block comment for the android/kotlin part of this PR :)

File: android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt
Comment (matamegger) [inline android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt:32]: I would suggest something like this:

```kotlin
if (json == null) return PlayerConfig()
            
val playerConfig = if (json.hasKey("licenseKey")) {
    PlayerConfig(key = json.getString("licenseKey"))
} else {
    PlayerConfig()
}
````

The suggestion also includes a more descriptive name for the "config" variable i.e. `playerConfig`

File: android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt
Comment (matamegger) [inline android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt:33]: If my code suggestion is applied, the `null`-check shouldn't be needed anymore.

File: android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt
Comment (matamegger) [inline android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt:40]: NIT (for all occurrences):
```suggestion
                if (playbackConfigJson?.hasKey("isAutoplayEnabled") == true) {
```

File: README.md
Comment (matamegger) [inline README.md:209]:
```
// Simply pass the `playbackConfig` property to `PlayerConfig` when instantiating a player.
```
This should be the `PlayerConfig` or? 🤔 
(cc @pogist)

Comment (pogist) [reply README.md:209]: yes, that should be `PlayerConfig`

File: android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt
Comment (jonathanm-tkf) [reply android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt:32]: Hi @matamegger @pogist 
At this point, I understand the guidelines but it loses a bit of sense as it is making the instance of PlayerConfig but does not match it to any variable. 

```java
     @JvmStatic
     fun toPlayerConfig(json: ReadableMap?): PlayerConfig {
        var config = PlayerConfig()
        if (json != null && json.hasKey("licenseKey")) {
            config = PlayerConfig(key = json.getString("licenseKey"))
        }
        if (json != null && json.hasKey("playbackConfig")) {
            var playbackConfigJson = json.getMap("playbackConfig")
            if (playbackConfigJson?.hasKey("isAutoplayEnabled") == true) {
                config.playbackConfig.isAutoplayEnabled = playbackConfigJson.getBoolean("isAutoplayEnabled")
            }
            /**
             * Specifies if the player should start muted.
             *
             * Default is `false`.
             */
            if(playbackConfigJson != null && playbackConfigJson.hasKey("isMuted")) {
                config.playbackConfig.isMuted = playbackConfigJson.getBoolean("isMuted")
            }
            /**
             * Specifies if time shifting (during live streaming) should be enabled.
             *
             * Default is `true`.
             */
            if(playbackConfigJson != null && playbackConfigJson.hasKey("isTimeShiftEnabled")) {
                config.playbackConfig.isTimeShiftEnabled = playbackConfigJson.getBoolean("isTimeShiftEnabled")
            }
        }
        return config
    }
```

The original idea is to be able to modify the values of ```config.playbackConfig.isAutoplayEnabled``` with the value received via parameter from RN, in the case that the config variable does not exist, it would not take this value as it is not specified, is this correct? or is this something I am missing?

**Review by jonathanm-tkf — COMMENTED** [review]
could you please check that everything is ok thank you very much @rolandkakonyi

File: android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt
Comment (matamegger) [reply android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt:32]: HI @jonathanm-tkf,
I don't quite understand your question.
I will try to elaborate on my suggested code snippet.

The function consists of many `if` statements at the moment. In each of the `if`s we check if `json` is `null`. In the case that `json` is indeed `null`, we basically do nothing and return `config`... which was initialized with a default `PlayerConfig`.
To avoid the many `null` checks I suggested the first line of my snippet:
```kotlin
if (json == null) return PlayerConfig()
```

Furthermore, `PlayerConfig.key` is a `val` and thus can not be changed after initializing the `PlayerConfig`. This means that we directly need to pass the `licenseKey` from the `ReadableMap` to the `PlayerConfig` constructor (if there is a `licenseKey` available in the `ReadableMap`).
The code in the PR already does this.

With my suggestion:
```kotlin
val playerConfig = if (json.hasKey("licenseKey")) {
    PlayerConfig(key = json.getString("licenseKey"))
} else {
    PlayerConfig()
}
```
We **always** assign a value for `playerConfig`. Depending on if there is a `licenseKey` in the `ReadableMap`, we either create a default `PlayerConfig` or a `PlayerConfig` with a player license key.

As mentioned in the first comment, I would like you to rename the current variable name `config` to `playerConfig`. i.e. there should be no local variable called `config` anymore :)

So if you replace:
```kotlin
var config = PlayerConfig()
if (json != null && json.hasKey("licenseKey")) {
    config = PlayerConfig(key = json.getString("licenseKey"))
}
```
with my suggestion, and of course every usage of `config` with `playerConfig`, everything should still work as intended by you. 🙂

File: android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt
Comment (matamegger) [inline android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt:42]: I think @rolandkakonyi suggestion referred to all the code comment blocks in the function(s) 🙂

File: android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt
Comment (matamegger) [inline android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt:50]: Please also remove this one.

File: ios/RCTConvert+BitmovinPlayer.swift
Comment (matamegger) [inline ios/RCTConvert+BitmovinPlayer.swift:23]: Also this comment is not needed, as well as the other block comments in this function.

**Review by matamegger — CHANGES_REQUESTED** [review]
I tried to provide you more details on my previous feedback.
Let me know if I missunderstood your question or something is unclear 🙂

File: android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt
Comment (jonathanm-tkf) [reply android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt]: Hi @matamegger thank you for your support, now I've caught it

apologies for the misunderstanding, I have updated the PR with the resolution as above and have taken the liberty of also removing some nullity checks as you had mentioned above, let me know if any other corrections are needed.

Regards.

File: ios/RCTConvert+BitmovinPlayer.swift
Comment (matamegger) [inline ios/RCTConvert+BitmovinPlayer.swift:26]: Please remove this block comment.

File: ios/RCTConvert+BitmovinPlayer.swift
Comment (matamegger) [inline ios/RCTConvert+BitmovinPlayer.swift:34]: Please remove this block comment.

File: ios/RCTConvert+BitmovinPlayer.swift
Comment (matamegger) [inline ios/RCTConvert+BitmovinPlayer.swift:42]: Please remove this block comment.

File: ios/RCTConvert+BitmovinPlayer.swift
Comment (matamegger) [inline ios/RCTConvert+BitmovinPlayer.swift:50]: Please remove this block comment.

**Review by rolandkakonyi — CHANGES_REQUESTED** [review]

**Review by matamegger — CHANGES_REQUESTED** [review]
I think you missed something from my previous feedback.

I explicitly added comments to the places I intended you to remove before. Sorry, if that was not entirely clear. I think after this we are good to go from my side 🙂 

EDIT: Started with my review before @rolandkakonyi and thus did't saw his comment. Sorry for repeating on the same thing.

**Review by rolandkakonyi — APPROVED** [review]

**Review by matamegger — APPROVED** [review]
🎉

**Pull request merged**
